### PR TITLE
TPD-1348 Divisions Dialog Rework

### PR DESCRIPTION
**Pull request by jason-rice**
lots of work done, looking great so far! lots more to do
[touchpointsoftware.atlassian.net/browse/TPD-1348](https://touchpointsoftware.atlassian.net/browse/TPD-1348)

### CODE REVIEW

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 1]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+                    var progName = item.Division.Program.IsNotNull() ? item.Division.Program.Name : "(no program)";
+                    var divName = item.Division.IsNotNull() ? item.Division.Name : "(no division)";
+                    var progDivString = progName + " - " + divName;
+                    output.Add(progDivString);
+                }
+            }
+            return Json(output, JsonRequestBehavior.AllowGet);
+        }
+
+        public JsonResult GetProgramsDivisionsForOrganizationArray(int orgId)
```
Comment (brondavies) [inline CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:141]: Why not use this method for the StringArray also?

File: CmsWeb/src/components/Modals/OrgDivAssignment.vue [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+                          v-on:click="moveDivisionUp(division, index)"
+                        ></i>
+                        <i
+                          class="fa fa-arrow-down"
+                          v-on:click="moveDivisionDown(division, index)"
+                        ></i>
+                        <i
+                          class="fa fa-trash"
+                          v-on:click="removeDivision(division)"
+                        ></i>
```
Comment (brondavies) [inline CmsWeb/src/components/Modals/OrgDivAssignment.vue:73]: These icons should each be wrapped in an `<a>` tag to make it function like a clickable element, preferably also with a `btn btn-link` class

Comment (brondavies) [reply CmsWeb/src/components/Modals/OrgDivAssignment.vue:73]: Hide the up arrow when it's the first in the list and hide the down arrow when it's the last item

File: CmsWeb/src/components/Modals/OrgDivAssignment.vue [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+              label="Name"
+            >
+            </MultiSelect>
+            <div class="table-responsive">
+              <table class="table">
+                <thead>
+                  <tr>
+                    <th style="width: 6%"></th>
+                    <th>Program : Division</th>
+                    <th style="width: 12%"></th>
```
Comment (brondavies) [inline CmsWeb/src/components/Modals/OrgDivAssignment.vue:48]: Use a fixed `px` width on the first and last column

**Review by brondavies — CHANGES_REQUESTED** [review]

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 4]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+        public JsonResult GetProgramsDivisionsForOrganizationStringArray(int orgId)
+        {
+            var divOrg = (from x in CurrentDatabase.DivOrgs where x.OrgId == orgId select x).ToList();
+            var output = new List<string>();
+            foreach (var item in divOrg)
+            {
+                if (item.Id.IsNotNull() && item.Id == 1)
+                {
+                    var progName = item.Division.Program.IsNotNull() ? item.Division.Program.Name : "(no program)";
+                    var divName = item.Division.IsNotNull() ? item.Division.Name : "(no division)";
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:123]: changed

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 5]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+            foreach (var item in divOrg)
+            {
+                if (item.Id.IsNull())
+                {
+                    var progName = item.Division.Program.IsNotNull() ? item.Division.Program.Name : "(no program)";
+                    var divName = item.Division.IsNotNull() ? item.Division.Name : "(no division)";
+                    var progDivString = progName + " - " + divName;
+                    output.Add(progDivString);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:137]: changed this

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 6]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+        public string FormatName(Division div)
+        {
+            if (div.Program.IsNotNull())
+            {
+                return div.Program.Name + " : " + div.Name;
+            }
+            else
+            {
+                return div.Name;
+            }
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:162]: that's awesome! changed this

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 7]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+        {
+            if (divorg.Division.Program.IsNotNull())
+            {
+                return divorg.Division.Program.Name + " : " + divorg.Division.Name;
+            }
+            else
+            {
+                return divorg.Division.Name;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:175]: very clever! changed this

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 8]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+                }
+                else
+                {
+                    var newDivOrg = new DivOrg()
+                    {
+                        Id = null,
+                        DivId = orgDivsList[i].Id,
+                        OrgId = orgId
+                    };
+                    CurrentDatabase.DivOrgs.InsertOnSubmit(newDivOrg);
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:201]: So I tried this change, but it didn't work. I did clean up this block of code though!

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 9]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+                    CurrentDatabase.SubmitChanges();
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:202]: Great catch! thanks

File: CmsWeb/Areas/Org/Views/Org/Settings/Main.cshtml [hunk 10]
```diff
@@ -2,6 +2,7 @@
 @using CmsWeb.Areas.Org.Models
 @using UtilityExtensions
 @model OrgMain
+@*<link rel="stylesheet" href="https://unpkg.com/vue-multiselect@2.1.0/dist/vue-multiselect.min.css">*@
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Views/Org/Settings/Main.cshtml:5]: oops, fixed this

File: CmsWeb/src/components/Modals/OrgDivAssignment.vue [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+    onSelect(option) {
+      let division = {
+        Id: option.Id,
+        Name: option.Name,
+      };
+      if (this.divisionsChosenList.length === 0) {
+        this.divisionsChosenList = [division];
+      } else {
+        this.divisionsChosenList.push(division);
+      }
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/Modals/OrgDivAssignment.vue:165]: not 100% sure why, but using just push wasn't working, but now it is, so changed this too

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs [hunk 12]
```diff
@@ -98,6 +100,112 @@ private void InitExportToolbar(OrganizationModel m)
...
+            var orgDivs = (from x in CurrentDatabase.Divisions
+                           select new
+                           {
+                               Id = x.Id,
+                               Name = FormatName(x)
+                           }).OrderBy(x => x.Id).ToList();
+            return Json(orgDivs, JsonRequestBehavior.AllowGet);
+        }
+
+        public JsonResult GetProgramsDivisionsForOrganizationStringArray(int orgId)
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:114]: changed this

File: CmsWeb/Areas/Org/Controllers/Org/OrgController.cs (hunk 1 above)
Comment (jason-rice) [reply CmsWeb/Areas/Org/Controllers/Org/OrgController.cs:141]: yeah, good point, I changed this

File: CmsWeb/src/components/Modals/OrgDivAssignment.vue (hunk 2 above)
Comment (jason-rice) [reply CmsWeb/src/components/Modals/OrgDivAssignment.vue:73]: made these changes, good suggestions!

File: CmsWeb/src/components/Modals/OrgDivAssignment.vue (hunk 3 above)
Comment (jason-rice) [reply CmsWeb/src/components/Modals/OrgDivAssignment.vue]: changed this

**Pull request merged**